src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -936,7 +936,83 @@ public static bool IsPointerType(this TypeSymbol type)
 
         internal static int FixedBufferElementSizeInBytes(this TypeSymbol type)
         {
-            return type.SpecialType.FixedBufferElementSizeInBytes();
+            // Fixed size buffer must be of type bool, byte, char, short, int, long, sbyte, ushort, uint, ulong, float,
+            // double, or an explicitly sized struct containing only fields that are one of the supported types (including
+            // fixed size buffers of the supported types).
+
+            var specialType = type.SpecialType;
+
+            if ((specialType != SpecialType.None) || (type.TypeKind != TypeKind.Struct))
+            {
+                // We are either a special type (bool, byte, char, short, int, long, sbyte, ushort, uint, ulong, float,
+                // or double) and we will return the correct element size or we are not a special type and we are also
+                // not a struct, in which case we will return 0.
+                return specialType.FixedBufferElementSizeInBytes();
+            }
```

Comment:
My preference would be to drop the comment and make the code explicit for non-special type, non-structs:
```
if (type.TypeKind != TypeKind.Struct)
{
    return 0;
}
```